examples/min-platform/embedding/wasmtime-platform.h

```
@@ -0,0 +1,203 @@
+ * point.
+ *
+ * The argument here was originally passed to `wasmtime_setjmp` through its
+ * out-param.
+ *
+ * This function cannot return.
+ *
+ * This function may be invoked from the `wasmtime_trap_handler_t`
+ * configured by `wasmtime_init_traps`.
+ */
+extern void wasmtime_longjmp(const uint8_t *jmp_buf);
+
+/**
+ * Initializes trap-handling logic for this platform.
+ *
+ * Wasmtime's implementation of WebAssembly relies on the ability to catch
+ * signals/traps/etc. For example divide-by-zero may raise a machine
+ * exception. Out-of-bounds memory accesses may also raise a machine
+ * exception. This function is used to initialize trap handling.
+ *
+ * The `handler` provided is a function pointer to invoke whenever a trap
+ * is encountered. The `handler` is invoked whenever a trap is caught by
+ * the system.
+ */
+extern void wasmtime_init_traps(wasmtime_trap_handler_t handler);
+
+/**
+ * Attempts to create a new in-memory image of the `ptr`/`len` combo which
+ * can be mapped to virtual addresses in the future. The returned
+ * `wasmtime_memory_image` pointer can be `NULL` to indicate that an image
+ * cannot be created. The structure otherwise will later be deallocated
+ * with `wasmtime_memory_image_free` and `wasmtime_memory_image_map_at`
+ * will be used to map the image into new regions of the address space.
+ */
+extern struct wasmtime_memory_image *wasmtime_memory_image_new(const uint8_t *ptr, uintptr_t len);
+
+/**
+ * Maps the `image` provided to the virtual address at `addr` and `len`.
+ *
+ * This semantically should make it such that `addr` and `len` looks the
+ * same as the contents of what the memory image was first created with.
+ * The mappings of `addr` should be private and changes do not reflect back
+ * to `wasmtime_memory_image`.
+ *
+ * In effect this is to create a copy-on-write mapping at `addr`/`len`
+ * pointing back to the memory used by the image originally.
+ *
+ * Aborts the process on failure.
```

Why not return an error and have the rust side panic? That would allow recovering from the error. It would be pretty bad for a kernel to simply crash because of an allocation failure.

---

I'd be tempted to say the embedding should crash in an embedding-defined way if possible to get better stack frames/debugging/etc. This API isn't intended to be set in stone until the end of time so I'm imagining it can change in the future to returning a result. In the meantime though whether the embedder aborts or Wasmtime aborts seems six-to-one-half-dozen-or-the-other to me

---

The embedder could return an error bringing down just a single wasm instance without crashing the entire process.

---

That's a good point yeah, and it's also true that `Result` is already in all these locations, so you've convinced me and I'll remove the assumed aborts here